tests/tests/integration_tests.rs

```diff
@@ -1206,4 +1206,72 @@ async fn graceful_shutdown_works() {
 
 	// The pending calls should be answered before shutdown.
 	assert_eq!(res.await.unwrap(), calls_len);
+
+	// The server should be closed now.
+	assert!(client.request::<String, _>("sleep_10s", rpc_params!()).await.is_err());
+}
+
+#[tokio::test]
+async fn graceful_shutdown_http_works() {
+	init_logger();
+
+	let (tx, mut rx) = tokio::sync::mpsc::unbounded_channel();
+	let call_answered = Arc::new(AtomicBool::new(false));
+
+	let (handle, addr) = {
+		let server = ServerBuilder::default().build("127.0.0.1:0").with_default_timeout().await.unwrap().unwrap();
+
+		let mut module = RpcModule::new((tx, call_answered.clone()));
+
+		module
+			.register_async_method("sleep_10s", |_, mut ctx| async move {
+				let ctx = Arc::make_mut(&mut ctx);
+				let _ = ctx.0.send(());
+				tokio::time::sleep(Duration::from_secs(10)).await;
+				ctx.1.store(true, std::sync::atomic::Ordering::SeqCst);
+				"ok"
+			})
+			.unwrap();
+		let addr = server.local_addr().unwrap();
+
+		(server.start(module).unwrap(), addr)
+	};
+
+	let client = HttpClientBuilder::default().build(format!("http://{addr}")).unwrap();
+
+	let mut calls: FuturesUnordered<_> = (0..10)
+		.map(|_| {
+			let c = client.clone();
+			async move { c.request::<String, _>("sleep_10s", rpc_params!()).await }
+		})
+		.collect();
+
+	let calls_len = calls.len();
+
+	let res = tokio::spawn(async move {
+		let mut c = 0;
+		while let Some(Ok(_)) = calls.next().await {
+			c += 1;
+		}
+		c
+	});
+
+	// All calls has been received by server => then stop.
+	for _ in 0..calls_len {
+		rx.recv().await.unwrap();
+	}
+
+	// Assert that no calls have been answered yet
+	// The assumption is that the server should be able to read 10 messages < 10s.
+	assert!(!call_answered.load(std::sync::atomic::Ordering::SeqCst));
+
+	// Stop the server.
+	handle.stop().unwrap();
+	handle.stopped().await;
+
+	// The pending calls should be answered before shutdown.
+	assert_eq!(res.await.unwrap(), 10);
```

Comment:
So If I understand right, the core of this test is that it makes some requests to the server which will take time to respond, and then it asks to stop the server (after all calls have definitely made it to the server but before they will have resolved) and all calls should resolve before the server stops?

Sounds good to me offhand! I assume that without your fix above this fails?

I wonder whether some comment might make it easier to understand the aim, like:
```rust
/// - Make 10 calls to the server.
/// - Once they have all reached the server but before they have responded, stop the server.
/// - All calls should be responded to before the server shuts down.
```

Out of interest what happens if, while it's shutting down (ie while waiting for calls to be responded to), you make some new calls? I'd expect any new calls to to rejected or something, but would they slow down the shutting down process?

Reply:
Yeah, I did one call after the `server.stopped returned` to ensure that the server is closed and doesn't accept new connections/calls

Reply:
Ah but what if new calls come in after you call `handle.stop()` but before `handle.stopped().await;` finishes?

Eg if the server is quite busy and new connections keep coming in; will it still shutdown?

Reply:
it's not tested here indeed but I think it should be possible to add

Reply:
> Sounds good to me offhand! I assume that without your fix above this fails?

Yes, it fails without my fix

Reply:
> it's not tested here indeed but I think it should be possible to add

It'd be good to check this somehow I think just to make sure that continuing new connections won't stop the server from ever shutting down :)

> Yes, it fails without my fix

Woop!